usaspending_api/references/v2/views/filter_tree/psc_filter_tree.py

```diff
@@ -0,0 +1,42 @@
+from usaspending_api.references.v2.views.filter_tree.filter_tree import UnlinkedNode, FilterTree
+from usaspending_api.references.models import PSC
+
+PSC_GROUPS = {"Research and Development": r"^A.$", "Service": r"^[b-z]$", "Product": r"^\d\d$"}
+
+
+class PSCFilterTree(FilterTree):
+    def raw_search(self, tiered_keys):
+        if len(tiered_keys) == 0:
+            return self._toptier_search()
+        elif len(tiered_keys) == 1:
+            return self._psc_from_group(tiered_keys[0])
+        else:
+            return self._psc_from_parent(tiered_keys[-1])
+
+    def _toptier_search(self):
+        return PSC_GROUPS.keys()
+
+    def _psc_from_group(self, group):
+        if group in PSC_GROUPS:
```

Comment:
Does the code allow non-keys to be provided to this method?

Reply:
Yes, just like it allows a user to search for children of a non existant PSC, NAICS or Federal Account. The most Restful thing might be to do a search a level above and return 404 on that condition, but I think the case could be made either way here, and that strategy seems to double computation time without too much benefit.

Reply:
Perhaps I didn't ask clearly. It appears the code is checking the parameter `group` is in the hard coded dict keys above:

```python
if group in PSC_GROUPS:
```
=>
```python
if group in("Research and Development", "Service", "Product")
```

I haven't dived into the code, but that seems odd to need this check this far down in the business logic. Shouldn't the group already be vetted to ensure it is an acceptable value?

Reply:
For a GET request there is no such thing as an unacceptable value, only an unacceptable format. Just like you can search for the snapshot of a non-existant FA (https://api.usaspending.gov/api/v2/federal_accounts/8899/fiscal_year_snapshot/2017/) or the Federal Accounts under a non-existant agency (https://dev-api.usaspending.gov/api/v2/references/filter_tree/tas/998/), there's nothing wrong with looking for PSC codes under a category that doesn't exist. There's an argument to be made that all of these should return 404 instead of an empty list, but that would be a change to make  for the filter trees as a whole, not special behavior that only happens at one level of one tree.
  The check happens here because it's the first area of code dedicated to this specific level of logic for this tree. And even when the check happens, it's to avoid a null pointer on the table, not to validate the request; the endpoint still returns an empty list if the group isn't found, there's just no appropriate regex to pass to the next function. Now that I think about it, I might just do a get_or_default call and smooth out this code a little.